### PR TITLE
Message validation (Part 1)

### PR DESCRIPTION
**Pull request by simone-kalbermatter**
This PR addresses part of the issue [#1375](https://github.com/dedis/popstellar/issues/1375) and improves message validation for the following messages (Data objects):
- CreateLao
- GreetLao
- UpdateLao
- StateLao

Bad examples of messages from the [protocol folder ](https://github.com/dedis/popstellar/tree/master/protocol/examples/messageData) were added to existing jsonValidationTest methods of the Data objects. The ones who were not rejected by the parsing (constraints were not enforced) are now checked at the construction of the Data objects. This was chosen as it is early in the process of receiving a message, before the handler gets the data.

An issue was that id's are parsed as a string and not base64 data, so they could have any format. Some things that are already checked at serialization include that keys are base64 data or that times are positive.

### CODE REVIEW

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/lao/CreateLao.java [hunk 1]
```diff
@@ -29,10 +31,18 @@ public class CreateLao extends Data {
...
-      String id, String name, long creation, PublicKey organizer, List<PublicKey> witnesses) {
+      @NonNull String id,
+      @NonNull String name,
+      long creation,
+      @NonNull PublicKey organizer,
+      @NonNull List<PublicKey> witnesses) {
+    // organizer and witnesses are checked to be base64 at deserialization
+    checkBase64(id, "id");
+    checkStringNotEmpty(name, "name");
+    checkValidTime(creation);
```
Comment (matteosz) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/lao/CreateLao.java:45]: Probably the only suggestion would be a more structured approach. A [template pattern](https://refactoring.guru/design-patterns/template-method) could be a solution. Having an abstract class with a factory method (could be `checkValidity`) which calls all the small methods that perform a punctual check and extend it with a class that overrides each of them in place too, so it's more extensible for the future

**Review by matteosz — COMMENTED** [review]
Very impressive! Good work 😄 
Just a small suggestion on the design choice

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/lao/CreateLao.java (hunk 1 above)
Comment (simone-kalbermatter) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/lao/CreateLao.java:45]: Sounds good, I will look into it!

**Review by matteosz — APPROVED** [review]
LGTM! Very good approach with the Builder 😄

**Review by Warhedgehog — CHANGES_REQUESTED** [review]
Impressive work, very well documented and tested. You should address the first code smell which is the addition of a private constructor to `MessageValidator` to prevent instantiations. I also have a small question. 
Keep up the good work 😄

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/lao/GreetLao.java [hunk 2]
```diff
@@ -33,28 +34,36 @@ public class GreetLao extends Data {
...
+      @NonNull String id,
       @NonNull String frontend,
       @NonNull String address,
-      List<PeerAddress> peers) {
-    // Peers can be empty and address can be the same
-    this.peers = new ArrayList<>(peers);
-    this.address = address;
+      @NonNull List<PeerAddress> peers) {
+    MessageValidator.verify().checkBase64(id, "id");
+    // Checking that the id matches the current lao id is done in the GreetLao handler
```
[the remark this answers is not in the document]
Comment (simone-kalbermatter) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/lao/GreetLao.java:52]: The message handler checks that the lao id matches the one that the sender requested the catchup for and got this GreetLao message in response. It is checking against a valid lao id that is already in the laoRepo, which I don't have access to here. In the constructor of the GreetLao, some information like the lao creation time is missing, so I could not add an additional check for the lao id here, but I still added a check that it is base64 encoded at least.

[the remark this answers is not in the document]
Comment (Warhedgehog) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/lao/GreetLao.java:52]: Very well !

**Pull request merged**